modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/seedlist/HistoricalSnapshotOrdinalEntries.scala

```diff
@@ -0,0 +1,76 @@
+package org.tessellation.node.shared.domain.seedlist
+
+import cats.MonadThrow
+import cats.syntax.all._
+import org.tessellation.schema.SnapshotOrdinal
+import derevo.cats.eqv
+import derevo.derive
+import eu.timepit.refined.auto._
+import eu.timepit.refined.numeric.NonNegative
+import eu.timepit.refined.refineV
+import fs2.data.csv.{CellDecoder, DecoderError}
+
+import scala.util.control.NoStackTrace
+
+@derive(eqv)
+case class HistoricalSnapshotOrdinalEntries(values: List[HistoricalSnapshotOrdinalEntry])
+
+case object InvalidRange extends NoStackTrace
+
+object HistoricalSnapshotOrdinalEntries {
+
+  implicit val historicalSnapshotOrdinalEntriesCellDecoder: CellDecoder[Option[HistoricalSnapshotOrdinalEntries]] =
+    CellDecoder.stringDecoder.emap { s =>
+      mkSnapshotOrdinalTimeline(s).attempt
+        .map(
+          _.leftMap(_ => new DecoderError("Unable to parse the ordinal timeline"))
+            .map(_.some)
+        )
+        .get
+    }
+
+  private val valueDelimiter = "-"
+  private val entryDelimiter = ";"
+
+  def toRange[G[_]: MonadThrow](str: String): G[(Long, Long)] =
+    str
+      .pure[G]
+      .map(_.split(valueDelimiter).map(_.trim))
+      .flatMap {
+        case Array(start, end, _*) =>
+          (start.toLong <= end.toLong)
+            .pure[G]
+            .ifM(
+              (start.toLong, end.toLong).pure[G],
+              MonadThrow[G].raiseError(InvalidRange)
+            )
+        case Array(start, _*) => (start.toLong, Long.MaxValue).pure[G]
+        case _ =>
+          MonadThrow[G].raiseError(InvalidRange)
+      }
+
+  def toRanges[G[_]: MonadThrow](str: String): G[List[(Long, Long)]] =
+    str.pure[G].map(_.split(entryDelimiter)).flatMap(_.toList.traverse(toRange[G]))
+
+  def mkSnapshotOrdinalTimeline[G[_]: MonadThrow](str: String): G[HistoricalSnapshotOrdinalEntries] =
+    toRanges(str)
+      .flatMap(
+        _.traverse {
+          case (a, b) =>
+            val c = refineV[NonNegative](a)
+            val d = refineV[NonNegative](b)
+
+            (c.isRight && d.isRight)
+              .pure[G]
+              .ifM(
+                HistoricalSnapshotOrdinalEntry(start = SnapshotOrdinal(c.toOption.get), end = SnapshotOrdinal(d.toOption.get)).pure[G],
```

Comment:
instead of checking `c.isRight` and then `c.toOption.get` you can use pattern match and unwrap it, like:
```
case (Right(start), Right(end)) => HistoricalSnapshotOrdinalEntry(start, end)
case _ => ... invalid range
```

Reply:
Reworked to

```scala
  def toRangeBoundary(str: String): Either[Throwable, RangeBoundary] = {
    val values = str
      .split(valueDelimiter)
      .map(_.trim)
      .map(
        _.toLongOption
          .flatMap(refineV[NonNegative](_).toOption)
      )

    values match {
      case Array(Some(start), Some(end), _*) =>
        Either.cond(
          start <= end,
          RangeBoundary(start = start, end = end),
          InvalidRange
        )
      case Array(Some(start), _*) =>
        RangeBoundary(
          start = start,
          end = Refined.unsafeApply[Long, NonNegative](Long.MaxValue)
        ).asRight
      case _ => InvalidRange.asLeft
    }
  }
```